File: dsu.py
class dsu:
    def __init__(self, N: int):
        self.parent = list(range(N))
        self.sz = [1]*N

    def root(self, x: int):
        if self.parent[x] == x:
            return x
        self.parent[x] = self.root(self.parent[x])
        return self.parent[x]

    def merge(self, x: int, y: int):
        x = self.root(x)
        y = self.root(y)
        if x != y:
            if self.sz[x] < self.sz[y]:
                x, y = y, x
            self.parent[y] = x
            self.sz[x] += self.sz[y]

    def same(self, x: int, y: int):
        return self.root(x) == self.root(y)

    def size(self, x: int):
        return self.sz[self.root(x)]

File: C++/graph/dsu.cpp
class dsu{
private:
    vector<int> parent;
    vector<int> sz;
public:
    dsu(int n){
        parent.resize(n);
        sz.assign(n, 1);
        for (int i = 0; i < n; i++) parent[i] = i;
    }

    int root(int x){
        if (parent[x] == x) return x;
        return parent[x] = root(parent[x]);
    }

    void merge(int x, int y){
        x = root(x);
        y = root(y);
        if (x == y) return;

        if (sz[x] < sz[y]) swap(x, y);
        parent[y] = x;
        sz[x] += sz[y];
    }

    bool same(int x, int y){
        return root(x) == root(y);
    }

    int size(int x){
        return sz[root(x)];
    }
};